rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,349 @@
+and others files in tf.lite directory.
+
+## User Benefit
+
+The main beneficiaries would be:
+
+a) ML community that get a clean way of defining of composite ops without having
+to commit to particular implementation (or whether to build new `tf_op`)
+
+b) Tool maintainers that wouldn't need to write complicated and brittle graph
+extraction code whose sole purpose is to reverse engineer tensorflow
+implementations. For example here are some tf-lite transformations that identifies
+composite ops like [prelu](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_prelu.cc),
+[lstm](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_lstm.cc), 
+[l2 pooling](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_l2_pool.cc), etc. Which essentially
+requires updates to lite whenever tensorflow modifies those implementations.
+
+_Blog post announcement_: "How to add new functions to tensorflow"
+
+_Subtitle_: "So you don't regret it later".
+
+## Design Proposal
+
+Tensorflow already provides a mechanism to define functions inside a graph
+called `tf.function`. In short, `tf.function` allows to define a subgraph with
+dedicated inputs and outputs, that then is substituted as a custom node into
+graphs. The gist of our proposal is to add a few standardized attributes to
+signal to the processing/executing library that this function implements a
+“standard” function. The consumer library then can choose to process it
+differently should they choose to. The important part is that we define a
+standard set of attributes that users and tooling can rely on.
+
+As a start, we propose to add “implements” and “reference” as two attributes
+defining a “standard” function.
+
+For example:
+
+```
+@function.library_function(
+    implements=“google.matmul_low_rank_matrix”,
+    reference=“arxiv.org/…”)
+def matmul_lowrank(x, y):
+   “Multiplies two low rank matrices. ”
+   # Basic implementation goes here.
+```
+
+Note we use a wrapper named library_function, that under-the-hood will create a
+standard tensorflow function with standard attributes set.
+
+Arguments:
+
+* `implements`:  we propose using namespaces with the top-level namespace
+being company name, with a few reserved namespaces indicating built-in ops, 
+if so desired.  This attributes indicates what function this subgraph implements,
+and provides a hint for downstream parsers that they can replace this with 
+their own implementation should they so choose. 
+
+* `reference`:  The goal for this attribute is to provide a stable reference for the function
+implementation details. This could be a reference to non-tensorflow library (e.g. numpy.linalg.special_matmul)
+or arxiv reference. The downstream tooling can use this value as an opaque cache to verify that 
+it implements the right flavor of the function. 
```

Comment:
The impact of this field on the downstream tool is blurry to me. The "implement" should be enough to replace a subgraph and this field seems redundant to me right now. 
(I can see it being useful for documentation purpose though)

Reply:
(this paragraph is obsolete, the field is dropped right?)